src/impl_methods.rs

```diff
@@ -1048,18 +1119,36 @@ impl<A, S, D> ArrayBase<S, D> where S: Data<Elem=A>, D: Dimension
     /// Return a mutable pointer to the first element in the array.
     #[inline(always)]
     pub fn as_mut_ptr(&mut self) -> *mut A
-        where S: DataMut
+        where S: DataRawMut
     {
-        self.ensure_unique(); // for RcArray
+        self.try_ensure_unique(); // for RcArray
         self.ptr
     }
 
+    /// Return a pointer array.
+    #[inline(always)]
+    pub fn as_array_ptr(&self) -> ArrayPtr<A, D> {
+        unsafe { ArrayPtr::new_(self.ptr, self.dim.clone(), self.strides.clone()) }
+    }
+
+    /// Return a mutable pointer array.
+    #[inline(always)]
+    pub fn as_array_ptr_mut(&mut self) -> ArrayPtrMut<A, D>
+        where S: DataRawMut
+    {
+        self.try_ensure_unique(); // for RcArray
```

Comment:
Nice PR.

I've been reading it and wondering if we need both a "const" and "mut" ArrayPtr type, we don't really have a real difference between raw pointer types. I guess this here is the real difference: We need this split if to "unshare" reference counted data before mutating.

The alternative would be to not have automatic unsharing with array ptrs, and it would be a footgun to us and consumers of the library?

Reply:
> I guess this here is the real difference: We need this split if to "unshare" reference counted data before mutating.

Yeah, I think that was the only real difference implementation-wise. There is a semantic advantage, too, to having separate types, and they help reduce mistakes. Rust has separate `*const T` and `*mut T` types, after all.

One other potential difference is [covariance](https://doc.rust-lang.org/nightly/nomicon/subtyping.html) over `A`. (This will come up once we switch `ArrayBase.ptr` to `NonNull`. `ArrayPtr` should be covariant over `A`; I'm not sure about `ArrayPtrMut`.)

> The alternative would be to not have automatic unsharing with array ptrs, and it would be a footgun to us and consumers of the library?

Well, we could have both `as_array_ptr` and `as_array_ptr_mut` return a common `ArrayPtr` type (where `as_array_ptr_mut` would perform the unsharing), but, as you said, it would be easy to forget to use the `_mut` one when necessary.